api/src/wfl/log.clj

```diff
@@ -0,0 +1,67 @@
+(ns wfl.log
+  "Logging for WFL"
+  (:require [clojure.data.json :as json]
+            [clojure.string    :as str])
+  (:import [java.time Instant]))
+
+(defn ^:private key-fn
+  "Preserve the namespace of `key` when qualified."
+  [key]
+  ((:key-fn json/default-write-options)
+   (if (qualified-keyword? key)
+     (str (namespace key) \/ (name key))
+     key)))
+
+(defprotocol Logger
+  "Log `edn` to `logger` as JSON."
+  (-write [logger edn]))
+
+(def disabled-logger
+  "A logger that does not log."
+  (reify Logger
+    (-write [_ _])))
+
+(def stdout-logger
+  "A logger to write to standard output"
+  (reify Logger
+    (-write [logger edn]
+      (-> edn
+          (json/write-str :key-fn key-fn :escape-slash false)
+          println))))
+
+(def ^:dynamic *logger*
+  "The logger now."
+  stdout-logger)
+
+(defmacro log
+  "Log `expression` with `severity`."
+  [severity expression]
```

Comment:
If your log function took an `expression & tail`, you can consolidate all the string concatenation here by
```clojure
(str/join " " (cons ~expression ~tail))
```